src/server/pps/server/pipeline_controller.go

```diff
@@ -65,23 +92,162 @@ var (
+	op.bumpLock.Lock()
+	defer op.bumpLock.Unlock()
+	op.bumpCnt++
+	switch t {
+	case writeOp:
+		op.writeBump = true
+		op.deleteBump = false
+	case deleteOp:
+		op.writeBump = false
+		op.deleteBump = true
+	}
+}
+
+func (op *pipelineOp) getAndUnsetBump() opType {
+	op.bumpLock.Lock()
+	defer op.bumpLock.Unlock()
+	if op.writeBump {
+		op.writeBump = false
+		return writeOp
+	}
+	if op.deleteBump {
+		op.deleteBump = false
+		return deleteOp
+	}
+	return noOp
+}
+
+// Start calls step for the given pipeline in a backoff loop.
+// When it encounters a stepError, returned by most pipeline controller helper
+// functions, it uses the fields to decide whether to retry the step or,
+// if the retries have been exhausted, fail the pipeline.
+//
+// Other errors are simply logged and ignored, assuming that some future polling
+// of the pipeline will succeed.
+func (op *pipelineOp) Start(t opType, timestamp time.Time, withMasterLock func(func()), cleanup func()) {
+	switch t {
+	case writeOp:
+		var errCount int
+		var stepErr stepError
+		err := backoff.RetryNotify(func() error {
+			// Create/Modify/Delete pipeline resources as needed per new state
+			return op.step(timestamp)
+		}, backoff.NewExponentialBackOff(), func(err error, d time.Duration) error {
+			errCount++
+			if errors.As(err, &stepErr) {
+				if stepErr.retry && errCount < maxErrCount {
+					log.Errorf("PPS master: error updating resources for pipeline %q: %v; retrying in %v",
+						op.pipelineInfo.Pipeline, err, d)
+					return nil
+				}
+			}
+			return errors.Wrapf(err, "could not update resource for pipeline %q", op.pipelineInfo.Pipeline)
+		})
+
+		// we've given up on the step, check if the error indicated that the pipeline should fail
+		if err != nil && errors.As(err, &stepErr) && stepErr.failPipeline {
+			failError := op.setPipelineFailure(fmt.Sprintf("could not update resources after %d attempts: %v", errCount, err))
+			if failError != nil {
+				log.Errorf("PPS master: error creating a pipelineOp for pipeline '%s': %v", op.pipelineInfo.Pipeline,
+					errors.Wrapf(failError, "error failing pipeline %q (%v)", op.pipelineInfo.Pipeline, err))
+			}
+			log.Errorf("PPS master: error creating a pipelineOp for pipeline '%s': %v", op.pipelineInfo.Pipeline,
+				errors.Wrapf(err, "failing pipeline %q", op.pipelineInfo.Pipeline))
+		}
+
+	case deleteOp:
+		err := op.deletePipelineResources()
+		log.Errorf("PPS master: error deleting pipelineOp resources for pipeline '%s': %v", op.pipelineInfo.Pipeline,
+			errors.Wrapf(err, "failing pipeline %q", op.pipelineInfo.Pipeline))
+	}
+
+	// when checking for a bump, we must first acquire the master lock followed by the op.BumpLock
+	nextOt := noOp
+	withMasterLock(func() {
+		switch op.getAndUnsetBump() {
+		case writeOp:
+			nextOt = writeOp
+		case deleteOp:
+			nextOt = deleteOp
+		case noOp:
+			cleanup()
+		}
+	})
+	if nextOt != noOp {
+		// this should be tail recursive?
+		op.Start(nextOt, timestamp, withMasterLock, cleanup)
```

Comment:
just a callout that I'm assuming this will get executed tail-recursively, avoiding technical downfalls with recursion here